snuba/datasets/entities/factory.py

```diff
@@ -11,69 +11,66 @@ class InvalidEntityError(SerializableException):
-    from snuba.datasets.entities.metrics import (
+    from snuba.datasets.entities.events import EventsEntity  # noqa: F401
+    from snuba.datasets.entities.metrics import (  # noqa: F401
         MetricsCountersEntity,
         MetricsDistributionsEntity,
         MetricsSetsEntity,
         OrgMetricsCountersEntity,
     )
-    from snuba.datasets.entities.outcomes import OutcomesEntity
-    from snuba.datasets.entities.outcomes_raw import OutcomesRawEntity
-    from snuba.datasets.entities.profiles import ProfilesEntity
-    from snuba.datasets.entities.sessions import OrgSessionsEntity, SessionsEntity
-    from snuba.datasets.entities.transactions import TransactionsEntity
-
-    dev_entity_factories: MutableMapping[EntityKey, Callable[[], Entity]] = {}
+    from snuba.datasets.entities.outcomes import OutcomesEntity  # noqa: F401
+    from snuba.datasets.entities.outcomes_raw import OutcomesRawEntity  # noqa: F401
+    from snuba.datasets.entities.profiles import ProfilesEntity  # noqa: F401
+    from snuba.datasets.entities.sessions import (  # noqa: F401
+        OrgSessionsEntity,
+        SessionsEntity,
+    )
+    from snuba.datasets.entities.transactions import TransactionsEntity  # noqa: F401
 
-    entity_factories: MutableMapping[EntityKey, Callable[[], Entity]] = {
-        EntityKey.DISCOVER: DiscoverEntity,
-        EntityKey.EVENTS: EventsEntity,
-        EntityKey.GROUPASSIGNEE: GroupAssigneeEntity,
-        EntityKey.GROUPEDMESSAGES: GroupedMessageEntity,
-        EntityKey.OUTCOMES: OutcomesEntity,
-        EntityKey.OUTCOMES_RAW: OutcomesRawEntity,
-        EntityKey.SESSIONS: SessionsEntity,
-        EntityKey.ORG_SESSIONS: OrgSessionsEntity,
-        EntityKey.TRANSACTIONS: TransactionsEntity,
-        EntityKey.DISCOVER_TRANSACTIONS: DiscoverTransactionsEntity,
-        EntityKey.DISCOVER_EVENTS: DiscoverEventsEntity,
-        EntityKey.METRICS_SETS: MetricsSetsEntity,
-        EntityKey.METRICS_COUNTERS: MetricsCountersEntity,
-        EntityKey.ORG_METRICS_COUNTERS: OrgMetricsCountersEntity,
-        EntityKey.METRICS_DISTRIBUTIONS: MetricsDistributionsEntity,
-        EntityKey.PROFILES: ProfilesEntity,
-        **(dev_entity_factories if settings.ENABLE_DEV_FEATURES else {}),
-    }
 
-    try:
-        entity = ENTITY_IMPL[name] = entity_factories[name]()
-        ENTITY_NAME_LOOKUP[entity] = name
-    except KeyError as error:
-        raise InvalidEntityError(f"entity {name!r} does not exist") from error
+@functools.lru_cache(maxsize=420)
+def get_entity(name: EntityKey) -> Entity:
+    registered_entity = Entity.from_name(name.value)
+    if registered_entity is not None:
+        return cast(Entity, registered_entity())
+    import_entities()
+    registered_entity = Entity.from_name(name.value)
+    if registered_entity is not None:
+        return cast(Entity, registered_entity())
 
-    return entity
+    raise InvalidEntityError(f"entity {name!r} does not exist")
```

Comment:
I think this is still continuing on the wrong data model we have now where entities can exist outside of datasets. 
Entities were meant to exist only within a dataset as a dataset is supposed to be a sealed partition of the snuba data model 
https://getsentry.github.io/snuba/architecture/datamodel.html#datasets

Which makes me think a `get_entity` method should only exist within a dataset and not by itself, which basically would change quite a bit the problem you are trying to solve. 

Now the constraint was not built from the beginning, so today it may be not trivial to add, though what about trying to fix the data model inconsistency first? Maybe the problem would not exist anymore.